common/types/ballot.go

```diff
@@ -273,12 +273,15 @@ func (b *Ballot) SignedBytes() []byte {
 
 // HashInnerBytes returns the hash of the InnerBallot.
 func (b *Ballot) HashInnerBytes() []byte {
-	h := hash.New()
+	h := hash.GetHasher()
 	_, err := codec.EncodeTo(h, &b.InnerBallot)
 	if err != nil {
 		log.With().Fatal("failed to encode InnerBallot for hashing", log.Err(err))
 	}
-	return h.Sum(nil)
+	sum := h.Sum(nil)
+	h.Reset()
+	hash.PutHasher(h)
+	return sum
 }
```

Comment:
I think it would be best to always put it back with `defer` to avoid leaking when the code is refactored in the future to have an early exit (for example due to error handling). Perhaps consider an API similar to the one in the `context` package? I.e. return the "free" function along with the hasher. Using it would look as follows:
```go
h, free := hash.GetHasher()
defer free()
```